rfcs/51-animation-composition.md

```diff
@@ -0,0 +1,450 @@
+animatable component and will not generically work with non-Rust components, it
+will also not avoid the userspace `unsafe` unless the parallel iteration is
+removed.
+
+### Relatonal `BoneBinding` as a Component
+Instead of using `BoneBindings` as a resource that is continually rebuilt every
+frame it's changed, `BoneBinding` could be a relational component, much like
+`Parent` or `Children`. This removes the need to scan the named hierarchy every
+frame, and allows trivial parallel iteration via `Query::par_for_each`. However:
+
+ - it does not avoid the need for userspace `unsafe`.
+ - maintaining the binding components is going to be a nightmare if there are
+   any hierarchy or name changes underneath an `AnimationGraph`.
+ - Commands require applying buffers in between maintanence and use in queries,
+   forces creation of a bottleneck sync point.
+ - Using it as a component means there will be secondary archetype moves if a
+   name or hierarchy changes, which increases archetype fragmentation.
+
+### Combining Binding and Sampling Steps: Parallelizing on AnimationGraph
+If we use the same `unsafe` tricks while parallelizing on `AnimationGraph`,
+there's a risk of aliased mutable access if the same entity's components are
+animated by two or more `AnimationGraph`s.
+
+
+## Prior art
+This proposal is largely inspired by Unity's [Playable][playable] API, which has
+a similar goal of building composable time-sequenced graphs for animation, audio,
+and game logic. Several other game engines have very similar APIs and features:
+
+ - Unreal has [AnimGraph][animgraph] for creating dynamic animations in
+   Blueprints.
+ - Godot has [animation trees][animation-trees] for creating dynamic animations in
+   Blueprints.
+
+Within the Rust ecosystem, [Fyrox][fyrox] has it's own animation system that is
+signifgantly higher level than this implementation. At time of writing, it
+supports blending, but does not support arbitrary property animation, only
+supporting transform based skinned mesh animation. Fyrox supports a high level
+state machine that can be used to transition between different states and blend
+trees, much like Unity's Mecanim animation system.
+
+The proposed API here doesn't purport or aim to directly replicate the features
+seen in these other engines, but provide the absolute bare minimum API so that
+engine developers or game developers can build them if they need to. It's
+entirely feasible to construct a higher level state machine animator like Unity's
+Mecanim or Fyrox's animation system on top of the graph detailed here.
+
+[playable]: https://docs.unity3d.com/Manual/Playables.html
+[animgraph]: https://docs.unrealengine.com/4.27/en-US/AnimatingObjects/SkeletalMeshAnimation/AnimBlueprints/AnimGraph/
+[animation-trees]: https://docs.godotengine.org/en/stable/tutorials/animation/animation_tree.html
+[fyrox]: https://fyrox-book.github.io/fyrox/animation/animation.html
+
+## Unresolved questions
```

Comment:
I think it's well inside the scope of this RFC. I think Unity and Unreal also have a similar feature for masking bones to apply a certain animation.

The solution to this would require us to do a per-bone weight blend during graph evaluation. Masked out bones would force the applied weight to zero. This can either be done with a bitmask from bone ID that is checked during blending.

Reply:
Makes sense to me! I'm cool with adding it or punting it, as long as the scenario is acknowledged / accounted for in the RFC.